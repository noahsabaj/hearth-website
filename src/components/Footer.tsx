import { GitHub, Download, MenuBook } from '@mui/icons-material';
import { Box, Container, Typography, Button, Grid } from '@mui/material';
import React, { memo } from 'react';
import { Link } from 'react-router-dom';

import { COLORS, SPACING, MISC } from '../constants';

const Footer: React.FC = memo(() => {
  return (
    <Box
      component='footer'
      role='contentinfo'
      sx={{
        py: SPACING.xxl / SPACING.unit, // 48 / 8 = 6
        borderTop: `1px solid ${COLORS.utils.border}`,
      }}
    >
      <Container>
        <Grid container spacing={4} justifyContent='center' sx={{ mb: SPACING.lg / SPACING.unit }}>
          <Grid item>
            <Button
              color='inherit'
              component='a'
              href={MISC.github.repoUrl}
              target='_blank'
              rel='noopener noreferrer'
              startIcon={<GitHub />}
              sx={{
                color: 'rgba(255,255,255,0.7)',
                '&:hover': { color: COLORS.primary.main },
              }}
            >
              GitHub
            </Button>
          </Grid>
          <Grid item>
            <Button
              color='inherit'
              component={Link}
              to='/docs'
              startIcon={<MenuBook />}
              sx={{
                color: 'rgba(255,255,255,0.7)',
                '&:hover': { color: COLORS.primary.main },
              }}
            >
              Documentation
            </Button>
          </Grid>
          <Grid item>
            <Button
              color='inherit'
              component={Link}
              to='/downloads'
              startIcon={<Download />}
              sx={{
                color: 'rgba(255,255,255,0.7)',
                '&:hover': { color: COLORS.primary.main },
              }}
            >
              Downloads
            </Button>
          </Grid>
        </Grid>
        <Typography variant='body2' color='text.secondary' align='center' component='p'>
          Â© 2025 Hearth Engine. Built with <span aria-label='fire'>ðŸ”¥</span> in Rust.
        </Typography>
      </Container>
    </Box>
  );
});

Footer.displayName = 'Footer';

export default Footer;
