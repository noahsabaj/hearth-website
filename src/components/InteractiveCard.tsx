import { Card, CardProps } from '@mui/material';
import { motion, useMotionValue, useSpring, useTransform } from 'framer-motion';
import React, { useRef, useCallback } from 'react';

interface InteractiveCardProps extends CardProps {
  children: React.ReactNode;
  tiltIntensity?: number;
  hoverScale?: number;
  shadowIntensity?: number;
  glowEffect?: boolean;
  springConfig?: {
    damping?: number;
    stiffness?: number;
  };
  disableInteraction?: boolean;
}

const InteractiveCard: React.FC<InteractiveCardProps> = ({
  children,
  tiltIntensity = 15,
  hoverScale = 1.02,
  shadowIntensity = 0.3,
  glowEffect = false,
  springConfig = { damping: 25, stiffness: 400 },
  disableInteraction = false,
  sx,
  onMouseEnter,
  onMouseLeave,
  ...props
}) => {
  const cardRef = useRef<HTMLDivElement>(null);

  // Motion values for mouse position
  const mouseX = useMotionValue(0);
  const mouseY = useMotionValue(0);

  // Spring animations for smooth movement
  const rotateX = useSpring(
    useTransform(mouseY, [-0.5, 0.5], [tiltIntensity, -tiltIntensity]),
    springConfig
  );
  const rotateY = useSpring(
    useTransform(mouseX, [-0.5, 0.5], [-tiltIntensity, tiltIntensity]),
    springConfig
  );

  // Scale and shadow animations
  const scale = useSpring(1, springConfig);
  const shadowScale = useSpring(1, springConfig);

  const handleMouseMove = useCallback(
    (event: React.MouseEvent<HTMLDivElement>) => {
      if (disableInteraction || !cardRef.current) return;

      const rect = cardRef.current.getBoundingClientRect();
      const centerX = rect.left + rect.width / 2;
      const centerY = rect.top + rect.height / 2;

      // Normalize mouse position (-0.5 to 0.5)
      const x = (event.clientX - centerX) / rect.width;
      const y = (event.clientY - centerY) / rect.height;

      mouseX.set(x);
      mouseY.set(y);
    },
    [disableInteraction, mouseX, mouseY]
  );

  const handleMouseEnterCard = useCallback(
    (event: React.MouseEvent<HTMLDivElement>) => {
      if (!disableInteraction) {
        scale.set(hoverScale);
        shadowScale.set(1.5);
      }

      if (onMouseEnter) {
        onMouseEnter(event);
      }
    },
    [disableInteraction, scale, shadowScale, hoverScale, onMouseEnter]
  );

  const handleMouseLeaveCard = useCallback(
    (event: React.MouseEvent<HTMLDivElement>) => {
      if (!disableInteraction) {
        mouseX.set(0);
        mouseY.set(0);
        scale.set(1);
        shadowScale.set(1);
      }

      if (onMouseLeave) {
        onMouseLeave(event);
      }
    },
    [disableInteraction, mouseX, mouseY, scale, shadowScale, onMouseLeave]
  );

  const cardStyles = {
    cursor: disableInteraction ? 'default' : 'pointer',
    transformStyle: 'preserve-3d' as const,
    ...sx,
  };

  const enhancedSx = {
    ...cardStyles,
    transition: 'box-shadow 0.3s ease',
    '&:hover': {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ...(sx as any)?.['&:hover'],
      boxShadow: glowEffect
        ? `0 20px 40px rgba(255, 69, 0, ${shadowIntensity}), 0 0 20px rgba(255, 69, 0, 0.1)`
        : `0 20px 40px rgba(0, 0, 0, ${shadowIntensity})`,
    },
  };

  if (disableInteraction) {
    return (
      <Card
        ref={cardRef}
        sx={enhancedSx}
        onMouseEnter={handleMouseEnterCard}
        onMouseLeave={handleMouseLeaveCard}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...props}
      >
        {children}
      </Card>
    );
  }

  return (
    <motion.div
      ref={cardRef}
      style={{
        rotateX,
        rotateY,
        scale,
        transformStyle: 'preserve-3d',
      }}
      onMouseMove={handleMouseMove}
      onMouseEnter={handleMouseEnterCard}
      onMouseLeave={handleMouseLeaveCard}
    >
      <Card
        sx={enhancedSx}
        component={motion.div}
        whileTap={{ scale: 0.98 }}
        transition={{ type: 'spring', ...springConfig }}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...props}
      >
        <motion.div
          style={{
            transform: 'translateZ(20px)',
            transformStyle: 'preserve-3d',
          }}
        >
          {children}
        </motion.div>
      </Card>
    </motion.div>
  );
};

// Enhanced icon button with hover effects
interface InteractiveIconButtonProps {
  children: React.ReactNode;
  onClick?: () => void;
  href?: string;
  target?: string;
  size?: 'small' | 'medium' | 'large';
  color?: string;
  hoverColor?: string;
  className?: string;
}

export const InteractiveIconButton: React.FC<InteractiveIconButtonProps> = ({
  children,
  onClick,
  href,
  target,
  size = 'medium',
  color = 'inherit',
  hoverColor = '#ff4500',
  className,
}) => {
  const sizeMap = {
    small: 32,
    medium: 40,
    large: 48,
  };

  const buttonSize = sizeMap[size];

  const buttonProps = href ? { as: 'a', href, target, onClick } : { as: 'button', onClick };

  return (
    <motion.div
      className={className}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...buttonProps}
      style={{
        display: 'inline-flex',
        alignItems: 'center',
        justifyContent: 'center',
        width: buttonSize,
        height: buttonSize,
        borderRadius: '50%',
        background: 'transparent',
        border: 'none',
        cursor: 'pointer',
        color,
        transition: 'color 0.3s ease',
      }}
      whileHover={{
        scale: 1.1,
        color: hoverColor,
        backgroundColor: 'rgba(255, 255, 255, 0.1)',
      }}
      whileTap={{ scale: 0.9 }}
      transition={{
        type: 'spring',
        stiffness: 400,
        damping: 25,
      }}
    >
      {children}
    </motion.div>
  );
};

export default InteractiveCard;
