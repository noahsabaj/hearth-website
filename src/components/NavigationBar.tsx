import {
  GitHub,
  Download,
  MenuBook,
  Home,
  Help,
  Keyboard,
  Engineering,
  Contrast,
  CollectionsBookmark,
  Speed,
  Newspaper,
} from '@mui/icons-material';
import {
  AppBar,
  Toolbar,
  Box,
  Typography,
  Button,
  IconButton,
  useTheme,
  useMediaQuery,
  Tooltip,
} from '@mui/material';
import React, { useRef, useEffect } from 'react';
import { Link, useLocation } from 'react-router-dom';

import SearchBar, { SearchBarRef } from './SearchBar';
import { COLORS, SPACING, LAYOUT, MISC, HIGH_CONTRAST_COLORS } from '../constants';
import { useKeyboardShortcutsContext } from '../contexts/KeyboardShortcutsContext';
import { useThemeContext } from '../contexts/ThemeContext';

interface NavigationBarProps {
  variant?: 'home' | 'docs' | 'downloads' | 'showcase' | 'faq';
}

const NavigationBar: React.FC<NavigationBarProps> = ({ variant = 'home' }) => {
  const theme = useTheme();
  const isMobile = useMediaQuery(theme.breakpoints.down('md'));
  const location = useLocation();
  const { showToast, setSearchFocusCallback } = useKeyboardShortcutsContext();
  const { highContrastMode, toggleHighContrastMode } = useThemeContext();
  const searchBarRef = useRef<SearchBarRef>(null);

  // Connect search bar focus to keyboard shortcuts
  useEffect(() => {
    setSearchFocusCallback(() => {
      searchBarRef.current?.focus();
    });
  }, [setSearchFocusCallback]);

  const isActive = (path: string) => {
    return location.pathname === path;
  };

  return (
    <AppBar
      position='fixed'
      sx={{
        background: COLORS.background.overlayLight,
        backdropFilter: LAYOUT.backdropFilter.blur,
        borderBottom: `1px solid ${theme.palette.divider}`,
      }}
      component='nav'
      role='navigation'
      aria-label='Main navigation'
    >
      <Toolbar sx={{ gap: 1 }}>
        {/* Logo and Title */}
        <Box sx={{ display: 'flex', alignItems: 'center', flexGrow: 1 }}>
          <Link
            to='/'
            style={{
              display: 'flex',
              alignItems: 'center',
              textDecoration: 'none',
              color: 'inherit',
            }}
          >
            <img
              src={MISC.logo.path}
              alt='Hearth Engine'
              style={{
                height: SPACING.navbar.logoHeight,
                marginRight: SPACING.navbar.logoMargin,
                backgroundColor: COLORS.utils.transparent,
              }}
            />
            {!isMobile && (
              <Typography variant='h6' sx={{ fontWeight: 700 }}>
                Hearth Engine
              </Typography>
            )}
          </Link>
        </Box>

        {/* Search Bar */}
        <SearchBar ref={searchBarRef} />

        {/* Navigation Links */}
        <Box sx={{ display: 'flex', alignItems: 'center', gap: 0.5 }}>
          {variant !== 'home' && !isMobile && (
            <Button
              color='inherit'
              component={Link}
              to='/'
              startIcon={<Home />}
              sx={{
                borderBottom: isActive('/') ? '2px solid' : '2px solid transparent',
                borderColor: theme.palette.primary.main,
                borderRadius: 0,
                '&:hover': {
                  borderBottom: '2px solid',
                  borderColor: theme.palette.primary.light,
                },
              }}
            >
              Home
            </Button>
          )}

          <Button
            color='inherit'
            component={Link}
            to='/docs'
            startIcon={!isMobile && <MenuBook />}
            sx={{
              borderBottom: isActive('/docs') ? '2px solid' : '2px solid transparent',
              borderColor: theme.palette.primary.main,
              borderRadius: 0,
              '&:hover': {
                borderBottom: '2px solid',
                borderColor: theme.palette.primary.light,
              },
            }}
          >
            {isMobile ? 'Docs' : 'Documentation'}
          </Button>

          <Button
            color='inherit'
            component={Link}
            to='/engine'
            startIcon={!isMobile && <Engineering />}
            sx={{
              borderBottom: isActive('/engine') ? '2px solid' : '2px solid transparent',
              borderColor: theme.palette.primary.main,
              borderRadius: 0,
              '&:hover': {
                borderBottom: '2px solid',
                borderColor: theme.palette.primary.light,
              },
            }}
          >
            Engine
          </Button>

          <Button
            color='inherit'
            component={Link}
            to='/downloads'
            startIcon={!isMobile && <Download />}
            sx={{
              borderBottom: isActive('/downloads') ? '2px solid' : '2px solid transparent',
              borderColor: theme.palette.primary.main,
              borderRadius: 0,
              '&:hover': {
                borderBottom: '2px solid',
                borderColor: theme.palette.primary.light,
              },
            }}
          >
            Downloads
          </Button>

          <Button
            color='inherit'
            component={Link}
            to='/showcase'
            startIcon={!isMobile && <CollectionsBookmark />}
            sx={{
              borderBottom: isActive('/showcase') ? '2px solid' : '2px solid transparent',
              borderColor: theme.palette.primary.main,
              borderRadius: 0,
              '&:hover': {
                borderBottom: '2px solid',
                borderColor: theme.palette.primary.light,
              },
            }}
          >
            Showcase
          </Button>

          <Button
            color='inherit'
            component={Link}
            to='/benchmarks'
            startIcon={!isMobile && <Speed />}
            sx={{
              borderBottom: isActive('/benchmarks') ? '2px solid' : '2px solid transparent',
              borderColor: theme.palette.primary.main,
              borderRadius: 0,
              '&:hover': {
                borderBottom: '2px solid',
                borderColor: theme.palette.primary.light,
              },
            }}
          >
            {isMobile ? 'Bench' : 'Benchmarks'}
          </Button>

          <Button
            color='inherit'
            component={Link}
            to='/updates'
            startIcon={!isMobile && <Newspaper />}
            sx={{
              borderBottom: isActive('/updates') ? '2px solid' : '2px solid transparent',
              borderColor: theme.palette.primary.main,
              borderRadius: 0,
              '&:hover': {
                borderBottom: '2px solid',
                borderColor: theme.palette.primary.light,
              },
            }}
          >
            Updates
          </Button>

          {!isMobile && (
            <Button
              color='inherit'
              component={Link}
              to='/faq'
              startIcon={<Help />}
              sx={{
                borderBottom: isActive('/faq') ? '2px solid' : '2px solid transparent',
                borderColor: theme.palette.primary.main,
                borderRadius: 0,
                '&:hover': {
                  borderBottom: '2px solid',
                  borderColor: theme.palette.primary.light,
                },
              }}
            >
              FAQ
            </Button>
          )}

          <IconButton
            color='inherit'
            onClick={() => {
              const event = new KeyboardEvent('keydown', { key: '?' });
              window.dispatchEvent(event);
              showToast('Keyboard shortcuts help opened');
            }}
            aria-label='Show keyboard shortcuts (?)'
            sx={{
              '&:hover': {
                backgroundColor: highContrastMode
                  ? HIGH_CONTRAST_COLORS.utils.shimmer
                  : COLORS.utils.shimmer,
              },
            }}
          >
            <Keyboard />
          </IconButton>

          <Tooltip
            title={highContrastMode ? 'Disable high contrast mode' : 'Enable high contrast mode'}
          >
            <IconButton
              color='inherit'
              onClick={() => {
                toggleHighContrastMode();
                showToast(
                  highContrastMode ? 'High contrast mode disabled' : 'High contrast mode enabled',
                );
              }}
              aria-label={
                highContrastMode ? 'Disable high contrast mode' : 'Enable high contrast mode'
              }
              sx={{
                '&:hover': {
                  backgroundColor: highContrastMode
                    ? HIGH_CONTRAST_COLORS.utils.shimmer
                    : COLORS.utils.shimmer,
                },
              }}
            >
              <Contrast />
            </IconButton>
          </Tooltip>

          <IconButton
            color='inherit'
            href={MISC.github.repoUrl}
            target='_blank'
            rel='noopener noreferrer'
            aria-label='View Hearth Engine on GitHub (opens in new tab)'
            sx={{
              '&:hover': {
                backgroundColor: COLORS.utils.shimmer,
              },
            }}
          >
            <GitHub />
          </IconButton>
        </Box>
      </Toolbar>
    </AppBar>
  );
};

export default NavigationBar;
