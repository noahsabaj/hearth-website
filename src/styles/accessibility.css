/* Accessibility CSS Utilities for Hearth Engine Website */

/* Screen reader only content */
.sr-only {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border: 0;
}

/* Show screen reader content on focus */
.sr-only-focusable:active,
.sr-only-focusable:focus {
  position: static;
  width: auto;
  height: auto;
  overflow: visible;
  clip: auto;
  white-space: normal;
}

/* Skip links */
.skip-link {
  position: absolute;
  top: -40px;
  left: 6px;
  background: #ff4500;
  color: white;
  padding: 8px;
  text-decoration: none;
  border-radius: 0 0 4px 4px;
  font-weight: 600;
  z-index: 9999;
  transition: top 0.3s ease;
}

.skip-link:focus {
  top: 0;
}

/* High contrast mode support */
@media (prefers-contrast: high) {
  :root {
    --primary-color: #ff6b00;
    --text-color: #ffffff;
    --background-color: #000000;
    --border-color: #ffffff;
  }

  .MuiButton-root {
    border: 2px solid currentColor !important;
  }

  .MuiCard-root {
    border: 2px solid rgba(255, 255, 255, 0.5) !important;
  }
}

/* Reduced motion preferences */
@media (prefers-reduced-motion: reduce) {
  *,
  *::before,
  *::after {
    animation-duration: 0.01ms !important;
    animation-iteration-count: 1 !important;
    transition-duration: 0.01ms !important;
    scroll-behavior: auto !important;
  }

  /* Disable all framer-motion animations */
  .motion-safe {
    transform: none !important;
    animation: none !important;
  }
}

/* Enhanced focus indicators */
:focus-visible {
  outline: 3px solid #ff4500 !important;
  outline-offset: 2px !important;
  border-radius: 4px;
}

/* Remove default focus styles only when :focus-visible is supported */
:focus:not(:focus-visible) {
  outline: none;
}

/* Improved focus for interactive elements */
button:focus-visible,
a:focus-visible,
input:focus-visible,
textarea:focus-visible,
select:focus-visible,
[tabindex]:focus-visible {
  outline: 3px solid #ff4500;
  outline-offset: 2px;
  box-shadow: 0 0 0 1px rgba(255, 69, 0, 0.3);
}

/* Touch target sizing */
.touch-target {
  min-height: 44px;
  min-width: 44px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
}

/* Color blind friendly indicators */
.status-success::before {
  content: '✓ ';
  color: #4caf50;
}

.status-error::before {
  content: '✗ ';
  color: #f44336;
}

.status-warning::before {
  content: '⚠ ';
  color: #ff9800;
}

.status-info::before {
  content: 'ℹ ';
  color: #2196f3;
}

/* Enhanced readability */
.readable-content {
  line-height: 1.6;
  max-width: 70ch;
  margin: 0 auto;
}

.readable-content p {
  margin-bottom: 1em;
}

/* Dark mode specific accessibility */
@media (prefers-color-scheme: dark) {
  :root {
    color-scheme: dark;
  }

  /* Ensure sufficient contrast for text */
  body {
    color: #e0e0e0;
    background-color: #0a0a0a;
  }

  /* Enhance focus visibility in dark mode */
  :focus-visible {
    outline-color: #ff6b35;
    box-shadow: 0 0 0 1px rgba(255, 107, 53, 0.3);
  }
}

/* Print accessibility */
@media print {
  .no-print {
    display: none !important;
  }

  a[href]::after {
    content: ' (' attr(href) ')';
    font-size: 0.8em;
    color: #666;
  }

  a[href^='#']::after,
  a[href^='javascript:']::after {
    content: '';
  }
}

/* Large text preferences */
@media (min-resolution: 192dpi) {
  body {
    font-weight: 400;
  }
}

/* Animation control utility classes */
.respect-motion-preference {
  animation-duration: 0.5s;
  transition-duration: 0.3s;
}

@media (prefers-reduced-motion: reduce) {
  .respect-motion-preference {
    animation-duration: 0.01ms;
    transition-duration: 0.01ms;
  }
}

/* ARIA live region styling */
.live-region {
  position: absolute;
  left: -10000px;
  width: 1px;
  height: 1px;
  overflow: hidden;
}

/* Keyboard navigation helpers */
.keyboard-navigation:focus-within {
  outline: 2px solid #ff4500;
  outline-offset: 2px;
}

/* Content grouping for better navigation */
.content-group {
  margin-bottom: 2rem;
}

.content-group:last-child {
  margin-bottom: 0;
}

/* Error state accessibility */
.error-state {
  border: 2px solid #f44336;
  background-color: rgba(244, 67, 54, 0.1);
}

.error-state::before {
  content: 'Error: ';
  font-weight: bold;
  color: #f44336;
}

/* Success state accessibility */
.success-state {
  border: 2px solid #4caf50;
  background-color: rgba(76, 175, 80, 0.1);
}

.success-state::before {
  content: 'Success: ';
  font-weight: bold;
  color: #4caf50;
}
