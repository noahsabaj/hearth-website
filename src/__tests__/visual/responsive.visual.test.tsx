/**
 * Responsive Visual Regression Tests
 * Tests components across different breakpoints and devices
 */

import { ThemeProvider, createTheme } from '@mui/material/styles';
import { render } from '@testing-library/react';
import React from 'react';
import { HelmetProvider } from 'react-helmet-async';
import { BrowserRouter } from 'react-router-dom';

// Components to test
import Footer from '../../components/Footer';
import InteractiveCard from '../../components/InteractiveCard';
import NavigationBar from '../../components/NavigationBar';
import VoxelLoader from '../../components/VoxelLoader';
import Home from '../../pages/Home';

// Test theme
const testTheme = createTheme({
  palette: {
    mode: 'dark',
    primary: { main: '#ff4500' },
    secondary: { main: '#ff6b35' },
  },
});

const TestWrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => (
  <BrowserRouter>
    <HelmetProvider>
      <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
    </HelmetProvider>
  </BrowserRouter>
);

// Viewport configurations
const VIEWPORTS = {
  mobile: { width: 375, height: 667, name: 'iPhone SE' },
  tablet: { width: 768, height: 1024, name: 'iPad' },
  desktop: { width: 1920, height: 1080, name: 'Desktop' },
  ultrawide: { width: 2560, height: 1440, name: 'Ultrawide' },
};

// Mock window.matchMedia for responsive tests
const mockMatchMedia = (width: number) => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches:
        query.includes(`max-width: ${width}px`) || query.includes(`min-width: ${width - 1}px`),
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
};

// Mock ResizeObserver
global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}));

describe('Responsive Visual Tests', () => {
  beforeEach(() => {
    // Reset viewport
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: 1920,
    });
    Object.defineProperty(window, 'innerHeight', {
      writable: true,
      configurable: true,
      value: 1080,
    });
  });

  Object.entries(VIEWPORTS).forEach(([device, viewport]) => {
    describe(`${viewport.name} (${viewport.width}x${viewport.height})`, () => {
      beforeEach(() => {
        // Set viewport dimensions
        Object.defineProperty(window, 'innerWidth', {
          writable: true,
          configurable: true,
          value: viewport.width,
        });
        Object.defineProperty(window, 'innerHeight', {
          writable: true,
          configurable: true,
          value: viewport.height,
        });

        mockMatchMedia(viewport.width);
      });

      it('should render NavigationBar responsively', () => {
        const { container } = render(
          <TestWrapper>
            <NavigationBar />
          </TestWrapper>
        );

        expect(container.firstChild).toBeInTheDocument();

        // Check for mobile menu on small screens
        if (viewport.width < 768) {
          expect(container.querySelector('[data-testid="mobile-menu"]')).toBeInTheDocument();
        } else {
          expect(container.querySelector('[data-testid="desktop-nav"]')).toBeInTheDocument();
        }
      });

      it('should render Footer with responsive layout', () => {
        const { container } = render(
          <TestWrapper>
            <Footer />
          </TestWrapper>
        );

        expect(container.firstChild).toBeInTheDocument();

        // Footer should stack vertically on mobile
        const footer = container.querySelector('footer');
        expect(footer).toBeInTheDocument();
      });

      it('should render VoxelLoader at appropriate size', () => {
        const { container } = render(
          <TestWrapper>
            <VoxelLoader size={viewport.width < 768 ? 'small' : 'medium'} />
          </TestWrapper>
        );

        expect(container.firstChild).toBeInTheDocument();
      });

      it('should render InteractiveCard with responsive behavior', () => {
        const { container } = render(
          <TestWrapper>
            <InteractiveCard>
              <div>Test Content</div>
            </InteractiveCard>
          </TestWrapper>
        );

        expect(container).toMatchSnapshot(`interactive-card-${device}`);
      });

      it('should render Home page layout responsively', () => {
        const { container } = render(
          <TestWrapper>
            <Home />
          </TestWrapper>
        );

        expect(container.firstChild).toBeInTheDocument();

        // Hero section should be present
        expect(container.querySelector('[data-testid="hero-section"]')).toBeInTheDocument();

        // Features section should adapt to viewport
        expect(container.querySelector('[data-testid="features-section"]')).toBeInTheDocument();
      });
    });
  });

  describe('Cross-Device Consistency', () => {
    it('should maintain consistent branding across all viewports', () => {
      Object.values(VIEWPORTS).forEach(viewport => {
        // Set viewport
        Object.defineProperty(window, 'innerWidth', {
          writable: true,
          configurable: true,
          value: viewport.width,
        });

        mockMatchMedia(viewport.width);

        const { container } = render(
          <TestWrapper>
            <NavigationBar />
          </TestWrapper>
        );

        // Logo should always be present
        expect(container.querySelector('[data-testid="nav-logo"]')).toBeInTheDocument();
      });
    });

    it('should handle viewport transitions smoothly', () => {
      const { container, rerender } = render(
        <TestWrapper>
          <NavigationBar />
        </TestWrapper>
      );

      // Start desktop
      Object.defineProperty(window, 'innerWidth', { value: 1920, writable: true });
      mockMatchMedia(1920);
      rerender(
        <TestWrapper>
          <NavigationBar />
        </TestWrapper>
      );

      // Transition to mobile
      Object.defineProperty(window, 'innerWidth', { value: 375, writable: true });
      mockMatchMedia(375);
      rerender(
        <TestWrapper>
          <NavigationBar />
        </TestWrapper>
      );

      // Should not crash during transition
      expect(container.firstChild).toBeInTheDocument();
    });
  });

  describe('Touch vs Mouse Interactions', () => {
    it('should handle touch events on mobile viewports', () => {
      Object.defineProperty(window, 'innerWidth', { value: 375, writable: true });
      mockMatchMedia(375);

      const { container } = render(
        <TestWrapper>
          <InteractiveCard>
            <div>Touch Test</div>
          </InteractiveCard>
        </TestWrapper>
      );

      // Simulate touch capability
      Object.defineProperty(window, 'ontouchstart', { value: true, writable: true });

      expect(container.firstChild).toBeInTheDocument();
    });

    it('should handle mouse interactions on desktop viewports', () => {
      Object.defineProperty(window, 'innerWidth', { value: 1920, writable: true });
      mockMatchMedia(1920);

      const { container } = render(
        <TestWrapper>
          <InteractiveCard>
            <div>Mouse Test</div>
          </InteractiveCard>
        </TestWrapper>
      );

      // Remove touch capability
      Object.defineProperty(window, 'ontouchstart', { value: undefined, writable: true });

      expect(container.firstChild).toBeInTheDocument();
    });
  });

  describe('Performance on Different Viewports', () => {
    it('should render efficiently on mobile devices', () => {
      Object.defineProperty(window, 'innerWidth', { value: 375, writable: true });
      mockMatchMedia(375);

      const startTime = performance.now();

      render(
        <TestWrapper>
          <Home />
        </TestWrapper>
      );

      const endTime = performance.now();
      const renderTime = endTime - startTime;

      // Mobile rendering should be fast
      expect(renderTime).toBeLessThan(100);
    });

    it('should handle high-DPI displays correctly', () => {
      Object.defineProperty(window, 'devicePixelRatio', { value: 2, writable: true });

      const { container } = render(
        <TestWrapper>
          <VoxelLoader />
        </TestWrapper>
      );

      expect(container.firstChild).toBeInTheDocument();
    });
  });
});
