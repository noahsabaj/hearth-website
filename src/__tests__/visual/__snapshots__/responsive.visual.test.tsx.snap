// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Responsive Visual Tests Desktop (1920x1080) should render InteractiveCard with responsive behavior: interactive-card-desktop 1`] = `
<div>
  <div
    style="transform: none;"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-l22w6d-MuiPaper-root-MuiCard-root"
      tabindex="0"
    >
      <div
        style="transform: translateZ(20px);"
      >
        <div>
          Test Content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Responsive Visual Tests Ultrawide (2560x1440) should render InteractiveCard with responsive behavior: interactive-card-ultrawide 1`] = `
<div>
  <div
    style="transform: none;"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-l22w6d-MuiPaper-root-MuiCard-root"
      tabindex="0"
    >
      <div
        style="transform: translateZ(20px);"
      >
        <div>
          Test Content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Responsive Visual Tests iPad (768x1024) should render InteractiveCard with responsive behavior: interactive-card-tablet 1`] = `
<div>
  <div
    style="transform: none;"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-l22w6d-MuiPaper-root-MuiCard-root"
      tabindex="0"
    >
      <div
        style="transform: translateZ(20px);"
      >
        <div>
          Test Content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Responsive Visual Tests iPhone SE (375x667) should render InteractiveCard with responsive behavior: interactive-card-mobile 1`] = `
<div>
  <div
    style="transform: none;"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-l22w6d-MuiPaper-root-MuiCard-root"
      tabindex="0"
    >
      <div
        style="transform: translateZ(20px);"
      >
        <div>
          Test Content
        </div>
      </div>
    </div>
  </div>
</div>
`;
